src/mlpack/core/optimizers/sgd/sgd_impl.hpp

```diff
@@ -103,7 +103,8 @@ double SGD<UpdatePolicyType, DecayPolicyType>::Optimize(
     }
 
     // Find the effective batch size (the last batch may be smaller).
-    const size_t effectiveBatchSize = std::min(batchSize,
+    const size_t effectiveBatchSize = std::min(
```

Comment:
I think the idea here is to make sure that the ``batchSize`` does not change? But it looks like this only works if ``actualMaxIterations >= numFunctions >= batchSize`` right?

Reply:
I made the change here to handle the situation where, e.g., `batchSize` is 50 but `actualMaxIterations` is, say, 5.  The current code will still execute 50 iterations of SGD because it will choose `effectiveBatchSize = 50`.  So here to choose the effective batch size we want to take the minimum of three quantities:

 - the user-specified batch size (`batchSize`)
 - the number of points remaining until we have looked at all functions (`numFunctions - currentFunction`)
 - the number of iterations remaining until the user wants optimization to terminate (`actualMaxIterations - i`)

Let me know if I've overlooked something or if that's actually not a good idea.

Reply:
I see, this is fine for me, it might be good to write that down at some point.

Reply:
Also, it might be good to do the same for the other optimizer as well, I can do that.

Reply:
No problem, let me add comments and I'll go ahead and add it to the other optimizers too (it's an easy fix).